File: libraries/typescript/src/converters/dwg-import.ts
/**
 * DWG to IFCX importer -- pure TypeScript, no external dependencies.
 *
 * Uses the built-in DwgParser to read DWG binary files and converts
 * the parsed data into an IfcxDocument.
 */

import { IfcxDocument } from '../document.js';
import { DwgParser, type DwgFile, type DwgObject } from './dwg-parser.js';

type AnyDict = Record<string, any>;

export class DwgImporter {
  /** Import DWG from raw bytes. */
  static fromBytes(data: Uint8Array): IfcxDocument {
    const parser = new DwgParser();
    const dwg = parser.parse(data);
    return DwgImporter.convert(dwg);
  }

  // ------------------------------------------------------------------
  // Conversion
  // ------------------------------------------------------------------

  private static convert(dwg: DwgFile): IfcxDocument {
    const doc = new IfcxDocument();
    doc.header = DwgImporter.convertHeader(dwg) as any;
    doc.tables = DwgImporter.convertTables(dwg) as any;
    doc.blocks = DwgImporter.convertBlocks(dwg) as any;
    doc.entities = DwgImporter.convertEntities(dwg) as any;
    doc.objects = DwgImporter.convertObjects(dwg) as any;
    return doc;
  }

  // ------------------------------------------------------------------
  // Header
  // ------------------------------------------------------------------

  private static convertHeader(dwg: DwgFile): AnyDict {
    const header: AnyDict = {};
    const hv = dwg.headerVars;

    header['version'] = hv['$ACADVER'] ?? dwg.versionCode;

    let insunits = hv['$LUNITS'] ?? 0;
    if (typeof insunits === 'object') insunits = 0;
    const unitMap: Record<number, string> = {
      0: 'unitless', 1: 'scientific', 2: 'decimal',
      3: 'engineering', 4: 'architectural', 5: 'fractional',
    };

    let measurement: unknown = hv['$MEASUREMENT'] ?? 1;
    if (typeof measurement === 'object') measurement = 1;

    header['units'] = {
      linear: unitMap[Number(insunits)] ?? 'unitless',
      measurement: Number(measurement) === 1 ? 'metric' : 'imperial',
    };

    const ltscale = hv['$LTSCALE'];
    if (typeof ltscale === 'number') header['linetypeScale'] = ltscale;

    return header;
  }

  // ------------------------------------------------------------------
  // Tables
  // ------------------------------------------------------------------

  private static convertTables(dwg: DwgFile): AnyDict {
    const tables: AnyDict = {
      layers: {},
      linetypes: {},
      textStyles: {},
      dimStyles: {},
    };

    for (const obj of dwg.objects) {
      if (obj.typeName === 'LAYER') {
        const name = String(obj.data['name'] ?? '');
        if (!name) continue;
        const props: AnyDict = {};
        if ('color' in obj.data) props['color'] = obj.data['color'];
        if ('frozen' in obj.data) props['frozen'] = obj.data['frozen'];
        if ('off' in obj.data) props['off'] = obj.data['off'];
        if ('locked' in obj.data) props['locked'] = obj.data['locked'];
        tables['layers'][name] = props;
      } else if (obj.typeName === 'STYLE') {
        const name = String(obj.data['name'] ?? '');
        if (!name) continue;
        const props: AnyDict = {};
        if ('fontName' in obj.data) props['fontFamily'] = obj.data['fontName'];
        if (obj.data['fixedHeight']) props['height'] = obj.data['fixedHeight'];
        if ('widthFactor' in obj.data) props['widthFactor'] = obj.data['widthFactor'];
        tables['textStyles'][name] = props;
      } else if (obj.typeName === 'LTYPE') {
        const name = String(obj.data['name'] ?? '');
        if (!name || name === 'ByBlock' || name === 'ByLayer' || name === 'Continuous') continue;
        const props: AnyDict = {};
        if ('description' in obj.data) props['description'] = obj.data['description'];
        if ('patternLength' in obj.data) props['patternLength'] = obj.data['patternLength'];
        tables['linetypes'][name] = props;
      }
    }

    if (!('0' in tables['layers'])) tables['layers']['0'] = {};

    return tables;
  }

  // ------------------------------------------------------------------
  // Blocks
  // ------------------------------------------------------------------

  private static convertBlocks(dwg: DwgFile): AnyDict {
    const blocks: AnyDict = {};
    for (const obj of dwg.objects) {
      if (obj.typeName === 'BLOCK_HEADER') {
        const name = String(obj.data['name'] ?? '');
        if (!name) continue;
        if (name.startsWith('*Model_Space') || name.startsWith('*Paper_Space')) continue;
        blocks[name] = { name, basePoint: [0, 0, 0], entities: [] };
      }
    }
    return blocks;
  }

  // ------------------------------------------------------------------
  // Entities
  // ------------------------------------------------------------------

  private static convertEntities(dwg: DwgFile): AnyDict[] {
    const entities: AnyDict[] = [];
    for (const obj of dwg.objects) {
      if (!obj.isEntity) continue;
      const converted = DwgImporter.convertEntity(obj);
      if (converted) entities.push(converted);
    }
    return entities;
  }

  private static convertEntity(obj: DwgObject): AnyDict | null {
    const result: AnyDict = { ...obj.data };
    const etype = String(result['type'] ?? '');

    // Normalize handle to string
    if ('handle' in result) {
      result['handle'] = typeof result['handle'] === 'number'
        ? result['handle'].toString(16).toUpperCase()
        : String(result['handle']);
    }

    // Remove internal fields
    for (const key of Object.keys(result)) {
      if (key.startsWith('_')) delete result[key];
    }

    // Color: 256 = BYLAYER, 0 = BYBLOCK -> remove
    if (result['color'] === 0 || result['color'] === 256) delete result['color'];

    // Remove zero thickness
    if (result['thickness'] === 0.0) delete result['thickness'];

    // Remove default extrusion [0,0,1]
    const ext = result['extrusion'];
    if (Array.isArray(ext) && ext[0] === 0 && ext[1] === 0 && ext[2] === 1) {
      delete result['extrusion'];
    }

    // Remove internal/default fields
    delete result['entity_mode'];
    delete result['linetype_scale'];
    delete result['invisible'];

    const lw = result['lineweight'];
    if (lw != null && (lw === 29 || lw < 0)) delete result['lineweight'];

    return result;
  }

  // ------------------------------------------------------------------
  // Objects
  // ------------------------------------------------------------------

  private static convertObjects(dwg: DwgFile): AnyDict[] {
    const objects: AnyDict[] = [];
    for (const obj of dwg.objects) {
      if (obj.typeName === 'DICTIONARY') {
        const converted: AnyDict = {
          objectType: 'DICTIONARY',
          handle: obj.handle.toString(16).toUpperCase(),
        };
        if ('entries' in obj.data) {
          const entries = obj.data['entries'] as Record<string, unknown>;
          converted['entries'] = Object.fromEntries(
            Object.entries(entries).map(([k, v]) => [
              k, typeof v === 'number' ? v.toString(16).toUpperCase() : String(v)
            ])
          );
        }
        objects.push(converted);
      }
    }
    return objects;
  }
}

File: libraries/rust/src/converters/dwg_import.rs
//! DWG to IFCX importer -- pure Rust.
//!
//! Uses the built-in `DwgParser` to read DWG binary files and converts
//! the parsed data into an `IfcxDocument`.

use std::collections::HashMap;
use std::path::Path;

use crate::types::*;
use crate::{IfcxDocument, IfcxError};
use super::dwg_parser::{DwgFile, DwgParser};

/// Imports DWG binary files into IFCX documents.
pub struct DwgImporter;

impl DwgImporter {
    /// Import DWG from a file path.
    pub fn from_file(path: impl AsRef<Path>) -> Result<IfcxDocument, IfcxError> {
        let data = std::fs::read(path)?;
        Self::from_bytes(&data)
    }

    /// Import DWG from raw bytes.
    pub fn from_bytes(data: &[u8]) -> Result<IfcxDocument, IfcxError> {
        let mut parser = DwgParser::new();
        let dwg = parser.parse(data)?;
        Ok(Self::convert(&dwg))
    }

    // ------------------------------------------------------------------
    // Conversion
    // ------------------------------------------------------------------

    fn convert(dwg: &DwgFile) -> IfcxDocument {
        let mut doc = IfcxDocument::new();

        doc.file.header = Self::convert_header(dwg);
        doc.file.tables = Some(Self::convert_tables(dwg));
        doc.file.blocks = Some(Self::convert_blocks(dwg));
        doc.file.entities = Self::convert_entities(dwg);
        doc.file.objects = Some(Self::convert_objects(dwg));

        doc
    }

    // ------------------------------------------------------------------
    // Header
    // ------------------------------------------------------------------

    fn convert_header(dwg: &DwgFile) -> Header {
        let version = dwg.header_vars.get("$ACADVER")
            .and_then(|v| v.as_str())
            .unwrap_or(&dwg.version_code)
            .to_string();

        let lunits = dwg.header_vars.get("$LUNITS")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);

        let unit_map: HashMap<i64, &str> = [
            (0, "unitless"), (1, "scientific"), (2, "decimal"),
            (3, "engineering"), (4, "architectural"), (5, "fractional"),
        ].iter().cloned().collect();

        let measurement = dwg.header_vars.get("$MEASUREMENT")
            .and_then(|v| v.as_i64())
            .unwrap_or(1);

        let linear = unit_map.get(&lunits).unwrap_or(&"unitless").to_string();
        let meas_str = if measurement == 1 { "metric" } else { "imperial" };

        let linetype_scale = dwg.header_vars.get("$LTSCALE")
            .and_then(|v| v.as_f64());

        Header {
            version: Some(version),
            author: None,
            organization: None,
            application: None,
            units: Some(Units {
                linear: Some(linear),
                measurement: Some(meas_str.to_string()),
                extra: HashMap::new(),
            }),
            extents: None,
            limits: None,
            current_layer: None,
            linetype_scale,
            extra: HashMap::new(),
        }
    }

    // ------------------------------------------------------------------
    // Tables
    // ------------------------------------------------------------------

    fn convert_tables(dwg: &DwgFile) -> Tables {
        let mut layers: HashMap<String, Layer> = HashMap::new();
        let mut linetypes: HashMap<String, serde_json::Value> = HashMap::new();
        let mut text_styles: HashMap<String, serde_json::Value> = HashMap::new();

        for obj in &dwg.objects {
            match obj.type_name.as_str() {
                "LAYER" => {
                    let name = obj.data.get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();
                    if name.is_empty() { continue; }

                    let color = obj.data.get("color")
                        .and_then(|v| v.as_i64())
                        .map(|c| Color::Index(c as i32));
                    let frozen = obj.data.get("frozen").and_then(|v| v.as_bool());
                    let off = obj.data.get("off").and_then(|v| v.as_bool());
                    let locked = obj.data.get("locked").and_then(|v| v.as_bool());

                    layers.insert(name, Layer {
                        color, linetype: None, lineweight: None,
                        frozen, locked, off, plot: None,
                    });
                }
                "STYLE" => {
                    let name = obj.data.get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();
                    if name.is_empty() { continue; }

                    let mut props = serde_json::Map::new();
                    if let Some(font) = obj.data.get("fontName").and_then(|v| v.as_str()) {
                        props.insert("fontFamily".into(), serde_json::json!(font));
                    }
                    if let Some(h) = obj.data.get("fixedHeight").and_then(|v| v.as_f64()) {
                        if h > 0.0 {
                            props.insert("height".into(), serde_json::json!(h));
                        }
                    }
                    text_styles.insert(name, serde_json::Value::Object(props));
                }
                "LTYPE" => {
                    let name = obj.data.get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();
                    if name.is_empty() || name == "ByBlock" || name == "ByLayer" || name == "Continuous" {
                        continue;
                    }

                    let mut props = serde_json::Map::new();
                    if let Some(desc) = obj.data.get("description").and_then(|v| v.as_str()) {
                        props.insert("description".into(), serde_json::json!(desc));
                    }
                    linetypes.insert(name, serde_json::Value::Object(props));
                }
                _ => {}
            }
        }

        // Ensure layer "0" exists
        layers.entry("0".to_string()).or_insert_with(|| Layer {
            color: None, linetype: None, lineweight: None,
            frozen: None, locked: None, off: None, plot: None,
        });

        Tables {
            layers: Some(layers),
            linetypes: Some(linetypes),
            text_styles: Some(text_styles),
            dim_styles: Some(HashMap::new()),
            extra: HashMap::new(),
        }
    }

    // ------------------------------------------------------------------
    // Blocks
    // ------------------------------------------------------------------

    fn convert_blocks(dwg: &DwgFile) -> HashMap<String, BlockDefinition> {
        let mut blocks = HashMap::new();
        for obj in &dwg.objects {
            if obj.type_name == "BLOCK_HEADER" {
                let name = obj.data.get("name")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string();
                if name.is_empty() { continue; }
                if name.starts_with("*Model_Space") || name.starts_with("*Paper_Space") {
                    continue;
                }
                blocks.insert(name.clone(), BlockDefinition {
                    name,
                    base_point: Some([0.0, 0.0, 0.0]),
                    entities: Some(Vec::new()),
                    extra: HashMap::new(),
                });
            }
        }
        blocks
    }

    // ------------------------------------------------------------------
    // Entities
    // ------------------------------------------------------------------

    fn convert_entities(dwg: &DwgFile) -> Vec<Entity> {
        dwg.objects.iter()
            .filter(|obj| obj.is_entity)
            .filter_map(|obj| Self::convert_entity(obj))
            .collect()
    }

    fn convert_entity(obj: &super::dwg_parser::DwgObject) -> Option<Entity> {
        let mut properties: HashMap<String, serde_json::Value> = HashMap::new();

        for (key, value) in &obj.data {
            if key == "type" || key == "handle" { continue; }
            if key.starts_with('_') { continue; }
            properties.insert(key.clone(), value.clone());
        }

        // Handle normalization
        let handle = Some(format!("{:X}", obj.handle));

        // Normalize: remove color 0/256 (BYLAYER/BYBLOCK)
        if let Some(serde_json::Value::Number(n)) = properties.get("color") {
            if n.as_i64() == Some(0) || n.as_i64() == Some(256) {
                properties.remove("color");
            }
        }

        // Remove zero thickness
        if let Some(serde_json::Value::Number(n)) = properties.get("thickness") {
            if n.as_f64() == Some(0.0) {
                properties.remove("thickness");
            }
        }

        // Remove default extrusion [0,0,1]
        if let Some(serde_json::Value::Array(arr)) = properties.get("extrusion") {
            if arr.len() == 3
                && arr[0].as_f64() == Some(0.0)
                && arr[1].as_f64() == Some(0.0)
                && arr[2].as_f64() == Some(1.0)
            {
                properties.remove("extrusion");
            }
        }

        // Remove internal fields
        properties.remove("entity_mode");
        properties.remove("linetype_scale");
        properties.remove("invisible");
        if let Some(serde_json::Value::Number(n)) = properties.get("lineweight") {
            let lw = n.as_i64().unwrap_or(0);
            if lw == 29 || lw < 0 {
                properties.remove("lineweight");
            }
        }

        Some(Entity {
            entity_type: obj.type_name.clone(),
            handle,
            layer: None,
            properties,
        })
    }

    // ------------------------------------------------------------------
    // Objects
    // ------------------------------------------------------------------

    fn convert_objects(dwg: &DwgFile) -> Vec<serde_json::Value> {
        let mut objects = Vec::new();
        for obj in &dwg.objects {
            if obj.type_name == "DICTIONARY" {
                let handle = format!("{:X}", obj.handle);
                let mut converted = serde_json::json!({
                    "objectType": "DICTIONARY",
                    "handle": handle,
                });
                if let Some(entries) = obj.data.get("entries") {
                    converted["entries"] = entries.clone();
                }
                objects.push(converted);
            }
        }
        objects
    }
}
